lib/fog/orchestration/util/recursive_hot_file_loader.rb

```diff
@@ -0,0 +1,227 @@
+        end
+
+        private
+
+        # Retrive the content of a local or remote file.
+        #
+        # @param A local or remote uri.
+        #
+        # @raise ArgumentError if it's not a valid uri
+        #
+        # Protect open-uri from malign arguments like
+        #  - "|ls"
+        #  - multiline strings
+        def read_uri(uri_or_filename)
+          remote_schemes = %w[http https ftp]
+          Fog::Logger.debug("Opening #{uri_or_filename}")
+
+          begin
+            # Validate URI to protect from open-uri attacks.
+            url = URI(uri_or_filename)
+
+            # Remote schemes must contain an host.
+            raise ArgumentError if url.host.nil? && remote_schemes.include?(url.scheme)
+
+            # Encode URI with spaces.
+            uri_or_filename = URI.encode(URI.decode(URI(uri_or_filename).to_s))
+          rescue URI::InvalidURIError
+            raise ArgumentError, "Not a valid URI: #{uri_or_filename}"
+          end
+
+          # TODO: A future revision may implement a retry.
+          # TODO: A future revision may limit download size.
+          content = ''
+          # open-uri doesn't open "file:///" uris.
+          uri_or_filename = uri_or_filename.sub(/^file:/, "")
+
+          open(uri_or_filename) { |f| content = f.read }
+          content
+        end
+
+        # Return true if the file is an heat template, false otherwise.
+        def template_is_raw?(content)
+          htv = content.strip.index("heat_template_version")
+          # Tolerate some leading character in case of a json template.
+          htv && htv < 5
+        end
+
+        # Return true if it's an URI, false otherwise.
+        def template_is_url?(path)
+          normalise_file_path_to_url(path)
+          true
+        rescue ArgumentError, URI::InvalidURIError
+          false
+        end
+
+        # Return true if I should I process this this file.
+        #
+        # @param [String] An heat template key
+        #
+        def ignore_if(key, value)
+          return true if key != 'get_file' && key != 'type'
+
+          return true unless value.kind_of?(String)
+
+          return true if key == 'type' &&
+                         !value.end_with?('.yaml', '.template')
+
+          false
+        end
+
+        # Return true if I should inspect this yaml template branch.
+        def recurse_if(value)
+          value.kind_of?(Hash) || value.kind_of?(Array)
+        end
+
+        # Returns the string baseurl of the given url.
+        def base_url_for_url(url)
+          parsed = URI(url)
+          parsed_dir = File.dirname(parsed.path)
+          url_join(parsed, parsed_dir)
+        end
+
+        def normalise_file_path_to_url(path)
+          # Nothing to do on URIs
+          return path if URI(path).scheme
+
+          path = File.absolute_path(path)
+          url_join('file:/', path)
+        end
+
+        def deep_copy(item)
+          return item if item.kind_of?(String)
```

Comment:
better
```
return item unless item.kind_of?(Hash)
```

Reply:
while imho:
- that won't deep_copy arrays
- may result counter-intuitive during further development

I'm fine for changing if it's a merge-blocker ;)